Feat/onramp implementation

sundayonah opened the pull request:
### Description

This PR implements the **Onramp transaction flow** for Noblocks, enabling users to swap local fiat currencies (NGN, KES, TZS, UGX) for stablecoins (USDC, USDT, etc.) to fund their wallets within the Paycrest ecosystem.

**User Story:**
As a Noblocks user, I want to easily and securely swap my local fiat currency (NGN, KES, TZS, UGX) for stablecoins (USDC, USDT, etc.) on the app, so that I can conveniently acquire crypto assets and fund my wallet within the Paycrest ecosystem.

**Key Features Implemented:**

1. **Onramp Transaction Detection & Flow:**
   - Modified swap interface to detect onramp transactions when fiat is selected as "Send" currency and stablecoin as "Receive" currency
   - Added conditional logic throughout the transaction flow to handle both onramp and offramp directions
   - Implemented wallet address input/selection for onramp recipients

### References

- Design Reference: https://www.figma.com/design/acOLQvbDj0TCm2ZKfl4y7e/Noblocks-Web-App?node-id=7-131530&t=fFaJ5PcVru4OFxn6-4
- Closes #299

### Checklist

- [x] I have added documentation and tests for new/changed/fixed functionality in this PR
- [ ] All active GitHub checks for tests, formatting, and security are passing
- [x] The correct base branch is being used, if not `main`

**Files Changed:**
- `noblocks/app/pages/TransactionStatus.tsx` - Onramp support, beneficiary modal, status messages
- `noblocks/app/pages/TransactionPreview.tsx` - Currency symbols, onramp detection
- `noblocks/app/pages/MakePayment.tsx` - Payment verification, currency symbols, pending state
- `noblocks/app/components/recipient/addBeneficiaryModal.tsx` - New component for wallet beneficiary name input
- `noblocks/app/components/recipient/RecipientListItem.tsx` - Display name for wallet recipients
- `noblocks/app/api/v1/recipients/route.ts` - Name validation for wallet recipients
- `noblocks/app/utils.ts` - Added `getCurrencySymbol` function
- `noblocks/app/types.ts` - Updated `RecipientDetails` type
- `noblocks/supabase/migrations/create_saved_recipients.sql` - Added name field to saved_wallet_recipients


<img width="1464" height="942" alt="Screenshot 2025-12-24 093533" src="https://github.com/user-attachments/assets/8e48922d-3c9e-49d9-b2a3-fa6191f3ab4e" />
<img width="1464" height="1004" alt="Screenshot 2025-12-24 093517" src="https://github.com/user-attachments/assets/7b633ac9-c4b3-4e3e-9ed8-093583fc3125" />
<img width="1461" height="1075" alt="Screenshot 2025-12-24 093504" src="https://github.com/user-attachments/assets/45b622a8-723d-4d67-8d91-79490c4ec3d4" />
<img width="1466" height="937" alt="Screenshot 2025-12-24 093448" src="https://github.com/user-attachments/assets/4e5f268f-76ca-4649-a297-25e9159f7a38" />


By submitting a PR, I agree to Paycrest's [Contributor Code of Conduct](https://paycrest.notion.site/Contributor-Code-of-Conduct-1602482d45a2806bab75fd314b381f4c) and [Contribution Guide](https://paycrest.notion.site/Contribution-Guide-1602482d45a2809a8930e6ad565c906a).

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Wallet recipient support: save and manage crypto wallet addresses as beneficiaries with avatar display
  * Onramp functionality: toggle between sending fiat and receiving crypto
  * Payment account details display with expiry countdown timer
  * Refund account management for payment processing
  * Network-aware beneficiary search and filtering

* **UI Enhancements**
  * Improved beneficiary modals with wallet-specific options
  * ENS name resolution for wallet recipients
  * Mode toggle for onramp and offramp flows
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

sundayonah committed:
feat: enhance swap functionality in TransactionForm

- Simplified the stateProps object in MainPageContent for better readability.
- Introduced isSwapped state in TransactionForm to manage swap mode between token and currency.
- Updated calculations for amount sent and received based on swap mode.
- Enhanced dropdowns and button for swapping fields, improving user experience.

sundayonah committed:
feat: implement `onramp` support and enhance transaction flow

- Added support for wallet recipients in the transaction process, allowing users to send tokens directly to wallet addresses.
- Introduced new state management for onramp mode, differentiating between wallet and bank/mobile money transactions.
- Updated API interactions to handle wallet recipient data, including saving and migrating wallet recipients.
- Enhanced UI components to accommodate wallet-specific fields and improve user experience during recipient selection.
- Implemented ENS name resolution for wallet addresses to provide better visibility for users.
- Updated database schema to include a new table for saved wallet recipients, ensuring proper data handling and security.

sundayonah committed:
feat: enhance recipient management and transaction flow

- Added a `name` field to wallet recipient details for better identification.
- Implemented `getCurrencySymbol` utility to standardize currency symbol retrieval across components.
- Updated transaction handling to validate and include recipient names in API interactions.
- Enhanced UI components to display recipient names and improve user experience during transactions.
- Modified database schema to ensure recipient names are stored and managed correctly.

sundayonah committed:
feat: enhance recipient validation and UI components

- Introduced type predicates for recipient details to improve type safety in filtering bank/mobile money and wallet recipients.
- Added validation for wallet address format in the API, ensuring correct input before processing.
- Updated UI components to utilize new type predicates for better handling of recipient data.
- Enhanced error handling for missing required fields in recipient data during local storage migration.
- Improved styling in the AddBeneficiaryModal for better user experience.

sundayonah committed:
feat: enhance TransactionForm with onramp/offramp buttons

- Added Onramp and Offramp buttons to the TransactionForm for improved user interaction.
- Updated layout to better accommodate new buttons while maintaining UI consistency.
- Ensured walletAddress is cleared when switching to onramp mode for better state management.

sundayonah committed:
Merge branch 'main' into feat/onramp-implementation

sundayonah committed:
Merge branch 'feat/onramp-implementation' of https://github.com/paycrest/noblocks into feat/onramp-implementation

sundayonah committed:
feat: add refund account management and payment order endpoints

- Introduced refund account API endpoints for fetching and saving refund account details.
- Added support for creating and retrieving v2 payment orders.
- Updated types to include new refund account and payment order structures.
- Enhanced middleware to include new API routes for refund accounts and payment orders.
- Updated configuration to include aggregator sender API key.
- Modified components and pages to integrate new refund account functionality and payment order handling.

sundayonah committed:
chore: remove referral feature files from onramp branch

Referral feature belongs to its own branch and has not been merged.
Removing all referral-specific files to keep this branch focused on the onramp implementation.

Made-with: Cursor

sundayonah committed:
feat: enhance transaction handling and add mobile sheet view type

- Introduced a new MobileSheetView type for better management of mobile sheet states.
- Updated TransactionPreview to reset the refund account on currency changes.
- Improved error handling in TransactionStatus for expired sessions during transaction tracking.

Dprof-in-tech commented on app/api/v1/payment-orders/[id]/route.ts:
```diff
@@ -0,0 +1,99 @@
+
+      // const baseUrl = config.aggregatorUrl.replace(/\/+$/, "").replace(/\/v1$/i, "");
+      const baseUrl = "https://api.paycrest.io";
+      const url = `${baseUrl}/v2/sender/orders/${encodeURIComponent(id)}`;
```
@sundayonah

onahprosper replied:
@coderabbitai, has this been resolved?

Dprof-in-tech replied:
check again, whether it has been resolved

onahprosper reviewed (changes requested):
For every local currency order, there should be fees attached; this fee will be added to the amount the user needs to transfer.
`senderFeePercent` must be included in the param and should determine the amount the user will pay.
==================

The rate for onramp should consider the provider buy rate and not the sell rate.

onahprosper committed:
Merge branch 'main' into feat/onramp-implementation

sundayonah committed:
feat: enhance onramp functionality and transaction status handling

- Added support for connected wallet address in RecipientDetailsForm.
- Updated transaction status types to include "expired".
- Implemented mapping of aggregator order status to database status for onramp transactions.
- Enhanced error handling and status resolution for onramp payment orders.
- Improved transaction details display for onramp transactions in TransactionDetails and TransactionPreview components.
- Refactored payment instruction mapping for better clarity and functionality.

sundayonah committed:
feat: introduce OnrampPaymentInstructions type and refactor mapping logic

- Added OnrampPaymentInstructions type to define the structure for virtual account/bank transfer instructions.
- Implemented mapProviderAccountToInstructions function to convert V2FiatProviderAccountDTO to OnrampPaymentInstructions.
- Removed redundant onrampPaymentInstructions module and updated imports in relevant components.
- Enhanced transaction handling in MakePayment and TransactionDetails components to utilize the new mapping function.

onahprosper commented on app/pages/TransactionForm.tsx:
```diff
@@ -448,13 +484,96 @@ export const TransactionForm = ({
+          // Use location-based currency if available and supported, otherwise default to NGN
+          const defaultCurrency = locationCurrency &&
+            currencies.find(c => c.name === locationCurrency && !c.disabled)
+            ? locationCurrency
+            : "NGN";
+
+          setValue("currency", defaultCurrency, { shouldDirty: true });
+        } catch {
+          // Fallback to NGN if location detection fails
+          setValue("currency", "NGN", { shouldDirty: true });
+        }
+      }
+      // Only set default token if not already selected
+      if (!token && fetchedTokens.length > 0) {
+        const usdcToken = fetchedTokens.find((t) => t.symbol === "USDC");
+        const defaultToken = usdcToken?.symbol || fetchedTokens[0]?.symbol;
+        if (defaultToken) {
+          setValue("token", defaultToken, { shouldDirty: true });
+        }
+      }
+      // Clear any existing walletAddress to start fresh
+      if (!walletAddress) {
+        setValue("walletAddress", "", { shouldDirty: true });
+      }
+    } else {
+      // Switching back to offramp mode
+      if (!token && fetchedTokens.length > 0) {
+        const usdcToken = fetchedTokens.find((t) => t.symbol === "USDC");
+        const defaultToken = usdcToken?.symbol || fetchedTokens[0]?.symbol;
+        if (defaultToken) {
+          setValue("token", defaultToken, { shouldDirty: true });
+        }
+      }
+      // Clear walletAddress when switching to offramp mode
+      setValue("walletAddress", "", { shouldDirty: true });
+    }
+
+    // Reset rate to trigger recalculation
+    stateProps.setRate(0);
+  };
```
@coderabbitai, has this been resolved?

onahprosper commented on app/types.ts:
```diff
@@ -49,15 +51,25 @@ export type TransactionPreviewProps = {
-  institutionCode: string;
-  accountIdentifier: string;
-  type: "bank" | "mobile_money";
-};
+export type RecipientDetails =
+  | {
+    type: "wallet";
+    walletAddress: string;
+    name: string;
+  }
+  | {
+    type: "bank" | "mobile_money";
+    name: string;
+    institution: string;
+    institutionCode: string;
+    accountIdentifier: string;
+    walletAddress?: never;
+  };
```
@coderabbitai, has this been resolved?

sundayonah commented on app/pages/TransactionPreview.tsx:
```diff
@@ -106,8 +124,12 @@ export const TransactionPreview = ({
     recipientName,
     accountIdentifier,
     memo,
+    walletAddress,
   } = formValues;
 
+  const isOnramp = !!walletAddress;
+  const currencySymbol = currency ? getCurrencySymbol(currency) : "";
+
```
@coderabbitai
We’re keeping isOnramp = !!walletAddress for now. On-ramp in this app is tied to having a recipient wallet for crypto delivery, and isSwapped isn’t persisted on FormData, so pair-only logic would need schema or submit-path changes.

sundayonah committed:
refactor: improve addBeneficiary function and transaction status handling

- Enhanced the addBeneficiary function to return a boolean indicating success or failure, improving error handling and user feedback.
- Cleaned up transaction status type definitions for better readability.
- Updated UI elements to reflect changes in transaction status and actions based on the current state.

sundayonah committed:
Merge remote-tracking branch 'origin/main' into feat/onramp-implementation

sundayonah committed:
feat: add isSwapped state to transaction forms and update related logic

- Introduced isSwapped boolean to FormData type for tracking swap mode.
- Updated MainPageContent and TransactionForm components to manage isSwapped state.
- Adjusted rate fetching logic based on isSwapped to differentiate between buy and sell sides.
- Enhanced wallet address handling to conditionally set isSwapped based on user input.

sundayonah committed:
fix: conditionally render receipt button based on onramp status

sundayonah committed:
feat: enhance transaction status handling for onramp flows

- Updated transaction completion logic to differentiate between onramp and off-ramp flows.
- Introduced processingStartedAt state to track the start time of onramp processing.
- Adjusted success visuals and button visibility based on transaction status for onramp transactions.
- Improved time spent calculations to reflect onramp processing duration accurately.

sundayonah committed:
feat: implement onramp client payment session management

- Added functionality to track and manage the onramp client payment session expiration.
- Introduced `isOnrampClientPaymentSessionExpired` utility to determine session validity based on transaction details.
- Updated `TransactionDetails`, `TransactionList`, and `MakePayment` components to reflect session expiration in UI.
- Enhanced transaction status handling to display "expired" when applicable, improving user feedback during onramp transactions.

Dprof-in-tech commented on app/utils.ts:
```diff
@@ -1267,3 +1375,72 @@ export function calculateSenderFee(
+  try {
+    await navigator.clipboard.writeText(text);
+    toast.success(label ? `${label} copied to clipboard` : "Copied to clipboard");
+  } catch (error) {
```
<img width="1578" height="1224" alt="Image" src="https://github.com/user-attachments/assets/9aa13a62-ba07-430f-9cd6-d7002058a4fe" />

we already had copy to clipboard functions across different files, if you localize it here , you might want to change everything to use this localized one across components that have a copy function like the wallet details and all that.

Dprof-in-tech replied:
these things will be addressed in a new PR

sundayonah replied:
this will be in another pr, i will use the local function one across the app.

sundayonah committed:
fix: correct base URL configuration for payment orders API

- Updated the base URL for the payment orders API to use the configured aggregator URL instead of a hardcoded value, ensuring proper integration with the aggregator service.

Dprof-in-tech commented on app/types.ts:
```diff
@@ -322,8 +412,9 @@ export type TransactionStatus =
-  | "refunded";
-export type TransactionHistoryType = "swap" | "transfer";
+  | "refunded"
+  | "expired";
```
i thought we said users should not see expired status?

Dprof-in-tech replied:
resolved

Dprof-in-tech commented on app/types.ts:
```diff
@@ -189,6 +216,66 @@ type TxReceipt = {
+  message: string;
+  data: T;
+};
+
```
if we are using v2 types, shouldn't we deprecate the corresponding v1 types?

Dprof-in-tech commented on app/types.ts:
```diff
@@ -81,7 +106,8 @@ export type TransactionStatusType =
   | "refunding"
-  | "refunded";
+  | "refunded"
+  | "expired";
```
i thought users shouldn't see expired ?

sundayonah replied:
the status is shown to users when pay-in deadline has passed.

Dprof-in-tech replied:
oh okay, nice nice

Dprof-in-tech commented on app/context/TransactionsContext.tsx:
```diff
@@ -6,11 +6,15 @@ import React, {
+import type { TransactionHistory, OrderDetailsData } from "../types";
 import {
   fetchTransactions,
   fetchOrderDetails,
```
again, since we are using v2 payment types and api, shouldn't we just move the entire app to use v2?

sundayonah replied:
nice catch
lots of features are still using v1 eg:(token,institution,verify_account etc)
i dont want to make the changes here, want to raise new pr to fix that.
this pr already have many changes.

Dprof-in-tech commented on app/api/v1/recipients/route.ts:
```diff
@@ -124,9 +153,146 @@ export const POST = withRateLimit(async (request: NextRequest) => {
     });
 
     const body = await request.json();
-    const { name, institution, institutionCode, accountIdentifier, type, currency } =
+    const { name, institution, institutionCode, accountIdentifier, type, currency, walletAddress: walletAddressFromBody } =
       body;
 
+    // Handle wallet recipients (onramp)
+    if (type === "wallet") {
+      if (!walletAddressFromBody) {
+        trackApiError(
+          request,
+          "/api/v1/recipients",
+          "POST",
+          new Error("Missing required field: walletAddress"),
+          400,
+        );
+        return NextResponse.json(
+          {
+            success: false,
+            error: "Missing required field: walletAddress",
+          },
+          { status: 400 },
+        );
+      }
+
+      // Validate wallet address format
+      if (!isValidEvmAddressCaseInsensitive(walletAddressFromBody.trim())) {
```
@sundayonah

Dprof-in-tech commented on app/api/v1/recipients/route.ts:
```diff
@@ -42,20 +43,33 @@ export const GET = withRateLimit(async (request: NextRequest) => {
+    // Fetch bank/mobile_money recipients
+    const { data: bankRecipients, error: bankError } = await supabaseAdmin
       .from("saved_recipients")
       .select("*")
       .eq("normalized_wallet_address", walletAddress)
       .order("created_at", { ascending: false });
 
-    if (error) {
-      console.error("Supabase query error:", error);
-      throw error;
+    if (bankError) {
+      console.error("Supabase query error:", bankError);
+      throw bankError;
+    }
+
+    // Fetch wallet recipients
+    const { data: walletRecipients, error: walletError } = await supabaseAdmin
+      .from("saved_wallet_recipients")
+      .select("*")
+      .eq("normalized_wallet_address", walletAddress)
+      .order("created_at", { ascending: false });
```
@sundayonah

Dprof-in-tech reviewed (changes requested):
@sundayonah please check out some of these comments

Dprof-in-tech reviewed (dismissed):
Looks good to me.

onahprosper reviewed (changes requested):
build is failing @sundayonah

Dprof-in-tech commented:
> build is failing @sundayonah

i dont see the deployment on DO ??

Dprof-in-tech commented:
> > build is failing @sundayonah
> 
> i dont see the deployment on DO ??

ohhhh, you mean locally

sundayonah committed:
fix: ensure contentClassName is safely handled in AnimatedModal component

- Updated the AnimatedModal component to use a fallback for contentClassName, preventing potential undefined values from causing issues in the className concatenation.

sundayonah commented:
> build is failing @sundayonah

fixed

Dprof-in-tech reviewed (approved):
LGTM!
